### DISCUSSION TITLE
Training model with ONNX runtime

### DISCUSSION THREAD

**Discussion by tiberiusferreira** [microsoft/onnxruntime#5307] [Other Q&A] [answered]
Hello, thanks for building the ONNX runtime! 

I'm trying to create a simple model and train it using onnx runtime, but I can't find an equivalent of `sess = onnxruntime.InferenceSession(path_save)` for training, is it something planned? I saw mentions of plans on making the training more seamless with Pytorch, which is great, but what I personally like about ONNX is how framework agnostic it is.

The tutorials seems to be Pytorch focused, but I'm more interested in using a "standalone" ONNX model, such as shown [here](https://github.com/microsoft/onnxruntime/blob/master/orttraining/tools/mnist_model_builder/mnist_fc_builder.ipynb) with the appropriate training information embedded in it, is this use case going to be supported? 

Nevertheless I tried the [getting started tutorial](https://github.com/microsoft/onnxruntime-training-examples/tree/master/getting-started), but it used 
`from onnxruntime.training import ORTTrainer, optim` which raised a `ModuleNotFoundError:` in both the latest stable and nightly build installed using `pip3 install -i https://test.pypi.org/simple/ ort-nightly`.

**Comment by tianleiwu** [comment, marked as answer]
@edgchen1, could you help on this topic?

**Reply by edgchen1** [reply]
Hi @tiberiusferreira, at the moment we don't publish stable or nightly Python wheels with training enabled.
We do have a Docker image with training enabled: mcr.microsoft.com/azureml/onnxruntime-training:0.1-rc2-openmpi4.0-cuda10.2-cudnn7.6-nccl2.7.6
FYI, there will be a new Docker image released soon. We are in the process of validating it now. The getting-started sample seems to currently refer to a private pre-release version, but that should change soon.
Another option is to build from source. See https://github.com/microsoft/onnxruntime/blob/master/BUILD.md#training for some instructions - feel free to ask questions if you run into any issues there.

**Reply by tiberiusferreira** [reply]
Thanks, I ran into a problem. 😓 I tried to build with "./build.sh --enable_training --build_wheel" but then it errors 
```
[ 97%] Building CXX object CMakeFiles/onnxruntime_pybind11_state.dir/Users/tiberiodarferreira/onnx-git/onnxruntime/python/onnxruntime_pybind_mlvalue.cc.o
In file included from /Users/tiberiodarferreira/onnx-git/onnxruntime/python/onnxruntime_pybind_mlvalue.cc:4:
In file included from /Users/tiberiodarferreira/onnx-git/onnxruntime/python/onnxruntime_pybind_mlvalue.h:7:
In file included from /Users/tiberiodarferreira/onnx-git/build/Linux/Debug/pybind11/src/pybind11/include/pybind11/iostream.h:12:
In file included from /Users/tiberiodarferreira/onnx-git/build/Linux/Debug/pybind11/src/pybind11/include/pybind11/pybind11.h:44:
In file included from /Users/tiberiodarferreira/onnx-git/build/Linux/Debug/pybind11/src/pybind11/include/pybind11/attr.h:13:
/Users/tiberiodarferreira/onnx-git/build/Linux/Debug/pybind11/src/pybind11/include/pybind11/cast.h:2011:26: error: loop variable 'a' is
      always a copy because the range of type 'detail::args_proxy' does not return a reference [-Werror,-Wrange-loop-analysis]
        for (const auto &a : ap)
```
I'm on MacOS.

**Reply by tiberiusferreira** [reply]
Seems to be same as https://github.com/google/filament/issues/3066 introduced by clang 12

**Reply by edgchen1** [reply]
Hm, unfortunately I don't have experience with building on Mac. It looks like our Mac CI build uses clang 11 (AppleClang 11.0.0.11000033).
You can file an issue - perhaps someone else will be able to help more.
Is using a different version of clang an option? Or building within a Docker container?

**Reply by tiberiusferreira** [reply]
I'll try downgrading clang, thanks!

**Reply by tiberiusferreira** [reply]
It worked, thanks for the help!

**Comment by SherlockNoMad** [comment]
We don't have plan to have a mirrored interface like onnxruntime.TrainingSession(). We usually recommend defining the model in pytorch nn.module and use ORTTrainer to interact with it. But if you already have a .onnx model together with loss function exported, you can also use ORTTrainer to load it. 
see https://github.com/microsoft/onnxruntime/blob/master/orttraining/orttraining/python/training/orttrainer.py#L68

**Reply by tiberiusferreira** [reply]
Thank you. Just to make sure I understood, it will be possible to user whatever tool to export a model.onnx file and load it up on the ORTTrainer and train it without having to go through Pytorch or any other framework, right? 

Some context: having a .onnx file containing everything needed for training and inference is very good for deploying to devices and doing on-device learning, but if one has to bundle Pytorch or something similar to do the training, it quickly becomes unfeasible.

So my vision was being able to have ONNX runtimes in general not only do inference, but also training.

**Reply by SherlockNoMad** [reply]
We share the same vision! 

Could you please share a bit more for on-device learning? What's the scenario, what's the underlaying device?

**Reply by PTMMachineLearning** [reply]
@tiberiusferreira Hello I am on a Facebook Pytorch team reaching out to understand the status and if everything is now working?

**Reply by tiberiusferreira** [reply]
Yep, it is. Thank you for reaching out.

**Reply by PTMMachineLearning** [reply]
@tiberiusferreira Thank you for the update.  If there is interest in accelerated ML workflow, training or inference, on MacOS I would be interested in knowing more about the project.  Thanks again.
